File: robot-factory/typescript/src/Factory.ts
import type { CostBreakdown } from './CostBreakdown.js';
import { Money } from './Money.js';
import { PartNotAvailableError } from './PartNotAvailableError.js';
import type { PartQuote } from './PartQuote.js';
import type { PartSupplier } from './PartSupplier.js';
import type { PurchasedPart } from './PurchasedPart.js';
import type { RobotOrder } from './RobotOrder.js';

export class Factory {
  private readonly suppliers: PartSupplier[];

  constructor(suppliers: PartSupplier[]) {
    this.suppliers = [...suppliers];
  }

  costRobot(order: RobotOrder): CostBreakdown {
    order.validate();

    const cheapestParts: PartQuote[] = [];

    for (const [type, option] of order.parts) {
      const quotes = this.suppliers
        .map(s => s.getQuote(type, option))
        .filter((q): q is PartQuote => q !== null);

      if (quotes.length === 0) {
        throw new PartNotAvailableError(`Part not available: ${option}`);
      }

      const cheapest = quotes.reduce((best, q) =>
        q.price.lessThan(best.price) ? q : best);
      cheapestParts.push(cheapest);
    }

    const total = cheapestParts.reduce(
      (sum, q) => sum.plus(q.price), Money.zero);

    return { parts: cheapestParts, total };
  }

  purchaseRobot(order: RobotOrder): PurchasedPart[] {
    const breakdown = this.costRobot(order);

    return breakdown.parts.map(quote => {
      const supplier = this.suppliers.find(s => s.name === quote.supplierName)!;
      return supplier.purchase(quote.type, quote.option);
    });
  }
}

File: robot-factory/python/src/robot_factory/factory.py
from __future__ import annotations

from functools import reduce

from .cost_breakdown import CostBreakdown
from .money import Money
from .part_not_available_error import PartNotAvailableError
from .part_supplier import PartSupplier
from .purchased_part import PurchasedPart
from .robot_order import RobotOrder


class Factory:
    def __init__(self, suppliers: list[PartSupplier]) -> None:
        self._suppliers = list(suppliers)

    def cost_robot(self, order: RobotOrder) -> CostBreakdown:
        order.validate()

        cheapest_parts = []

        for part_type, option in order.parts.items():
            quotes = [
                q
                for s in self._suppliers
                if (q := s.get_quote(part_type, option)) is not None
            ]

            if not quotes:
                raise PartNotAvailableError(
                    f"Part not available: {option.value}"
                )

            cheapest_parts.append(min(quotes, key=lambda q: q.price))

        total = reduce(lambda s, q: s + q.price, cheapest_parts, Money.zero())
        return CostBreakdown(parts=cheapest_parts, total=total)

    def purchase_robot(self, order: RobotOrder) -> list[PurchasedPart]:
        breakdown = self.cost_robot(order)

        purchased = []
        for quote in breakdown.parts:
            supplier = next(
                s for s in self._suppliers if s.name == quote.supplier_name
            )
            purchased.append(supplier.purchase(quote.type, quote.option))

        return purchased
